### ISSUE TITLE
Add build with sanitizers to CI

### ISSUE THREAD

**Issue by bayandin** [neondatabase/neon#6053]
Let's add a build with sanitizers (asan, ubsan) to the CI pipeline and run tests on it. To not blow up the CI workflow time, let's focus on Postgres 16 first.

Things to keep in mind:
- Enable sanitizers with gcc
- Add sanitizers to debug build, measure impact on build time and regression test suite, if it affects build/test time significantly, then do not enable by default (and hide behind some env vars)

An example can be found in [this slack thread](https://neondb.slack.com/archives/C04DGM6SMTM/p1699631690950259?thread_ts=1699631659.728189&cid=C04DGM6SMTM).

**Comment by abhigets** [comment]
thanks

**Comment by a-masterov** [comment]
Hi Alexander, Team,
First, in the original message, AL described patches for use with Clang, not Gcc.
Fortunately, MacOS uses Clang even while we are running binary with the name gcc:
```
alex@Alexeys-MacBook-Pro neon % /usr/bin/gcc --version
Apple clang version 15.0.0 (clang-1500.3.9.4)
Target: arm64-apple-darwin23.4.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin`
```
On the other hand, the Clang version built in MacOS is 15, but the original message mentioned Clang 18.
E.g. Clang v15 issues a warning about '-Wno-cast-function-type-strict' directive.
So, I will go on the experiments in docker with the newer version of Clang.

**Comment by a-masterov** [comment]
So I have figured out the newer version of Clang on MacOSX
```
brew install llvm
```
But then I faced another problem
```
configure:4022: clang -O0 -g3  -fsanitize=address -fsanitize=undefined -fno-sanitize-recover -fno-sanitize=function -Wno-cast-function-type-strict -fsanitize=address -fsanitize=undefined -static-libsan conftest.c  
>&5
clang: error: static AddressSanitizer runtime is not supported on darwin
```
I tried using docker, but I am getting an error "killed with signal 9", probably due to memory limit, so I am trying virtualization.

**Comment by a-masterov** [comment]
[sanitizer01.txt](https://github.com/neondatabase/neon/files/15074944/sanitizer01.txt)
Here are memory leaks found by ASAN.

**Comment by bayandin** [comment]
Last Week:
- The work started
- Reviewed and updated prior patches
- Working on passing flags to Rust compiler

This Week:
- Create the initial version of a PR with sanitizers

**Comment by bayandin** [comment]
Last Week:
- A PR is created: https://github.com/neondatabase/neon/pull/10401

This Week:
- Review PR
- Review flaky tests found by sanitizers

**Comment by bayandin** [comment]
Last Week:
- Helped with https://github.com/neondatabase/neon/pull/10280 to get rid of libpq dep from storage controller
- Prepared a separate PR with fixes in C-part and Postgres for sanitizers
- Review of new flaky tests

This Week:
- Merge https://github.com/neondatabase/neon/pull/10473
- Review more flaky tests - create issues

**Comment by bayandin** [comment]
Blocked by https://github.com/neondatabase/neon/pull/10614

**Comment by bayandin** [comment]
It's done!
A small follow-up: https://github.com/neondatabase/neon/pull/10731
